File: src/main/kotlin/diff/Runner.kt
package com.revtekk.blaze.diff

import com.revtekk.blaze.common.Command
import com.revtekk.blaze.common.CommandArgumentValidationException
import com.revtekk.blaze.common.readLines
import java.io.File

/**
 * Command to compute the difference between two files
 *
 * @property args first file path, second file path
 * @throws CommandArgumentValidationException if insufficient arguments are provided to run the command
 *
 * Usage: diff <first_file> <second_file>
 */
class DiffCommand: Command {
    constructor(args: List<String>) : super(args) {
        if (args.size != 2) {
            throw CommandArgumentValidationException(
                "diff: insufficient arguments specified, expected 2, got ${args.size}")
        }
    }

    override fun execute() {
        val first = File(args[0])
        val second = File(args[1])

        val firstLines = readLines(first.path)
        val secondLines = readLines(second.path)

        val fileDiff = diff(firstLines, secondLines)
        printDiff(fileDiff)
    }
}

File: src/diff/runner.rs
use crate::command::{Command, CommandExecutionError};
use crate::diff::diff;
use crate::diff::printer::print_diff;
use crate::file::read_lines;

pub struct DiffCommand {
    pub first: String,
    pub second: String,
}

impl Command for DiffCommand {
    fn execute(&self) -> Result<(), CommandExecutionError> {
        let first_lines = match read_lines(&self.first) {
            Ok(lines) => lines,
            Err(err) => return Err(err),
        };
        let second_lines = match read_lines(&self.second) {
            Ok(lines) => lines,
            Err(err) => return Err(err),
        };

        let file_diff = diff(&first_lines, &second_lines);
        print_diff(file_diff);
        Ok(())
    }
}
